packages/editor/src/validators/document/index.ts

```diff
@@ -1,56 +1,126 @@
 import type { Editor } from '@tiptap/core';
+import type { Node } from 'prosemirror-model';
 import type { EditorSettings } from '@/types/settings.ts';
 import type { DocumentValidator, ValidationResult } from '@/types/validation.ts';
 import { documentValidations, validationSeverity } from '@/validators/constants.ts';
 import { getNodeBoundingBox } from '@/validators/helpers.ts';
 
 const documentValidators = new Map<string, DocumentValidator>();
 
-export const documentMustHaveCorrectHeadingOrder = (editor: Editor): ValidationResult | null => {
-  let errorHeadingLevel = 0;
-  let errorPrecedingHeadingLevel = 0;
-
+export const documentMustHaveCorrectHeadingOrder = (editor: Editor): ValidationResult[] => {
+  const errors: ValidationResult[] = [];
   let precedingHeadingLevel = 0;
-  let errorPosition: number | null = null;
+
   editor.$doc.node.descendants((node, pos) => {
     if (node.type.name === 'heading') {
       const headingLevel = node.attrs['level'];
       if (headingLevel > precedingHeadingLevel + 1) {
-        errorPosition = pos;
-        errorPrecedingHeadingLevel = precedingHeadingLevel;
-        errorHeadingLevel = headingLevel;
+        errors.push({
+          boundingBox: getNodeBoundingBox(editor, pos),
+          pos,
+          severity: validationSeverity.WARNING,
+          tipPayload: { headingLevel: headingLevel, precedingHeadingLevel: precedingHeadingLevel },
+        });
       }
       precedingHeadingLevel = headingLevel;
     }
   });
-  if (errorPosition) {
-    return {
-      boundingBox: getNodeBoundingBox(editor, errorPosition),
-      pos: errorPosition,
-      severity: validationSeverity.WARNING,
-      tipPayload: { headingLevel: errorHeadingLevel, precedingHeadingLevel: errorPrecedingHeadingLevel },
-    };
+  return errors;
+};
+
+const getParagraphLines = (node: Node): string[] => {
+  const lines: string[] = [];
+  let buffer = '';
+  node.content.forEach((child) => {
+    if (child.type.name === 'hardBreak') {
+      if (buffer.trim().length) {
+        lines.push(buffer);
+      }
+      buffer = '';
+    } else {
+      buffer += child.textContent;
+    }
+  });
+  if (buffer.trim().length) {
+    lines.push(buffer);
+  }
+  return lines;
+};
+
+export const documentMustHaveSemanticLists = (editor: Editor): ValidationResult[] => {
+  const errors: ValidationResult[] = [];
+  const $blocks: { node: Node; pos: number }[] = [];
+  editor.$doc.node.descendants((node: Node, pos: number) => {
+    if (node.type.isBlock) {
+      $blocks.push({ node, pos });
+    }
+  });
+
+  const orderedListIndicator = new RegExp(/^\d+[.)\]/ ]$/);
+  const unorderedListIndicator = new RegExp(/^\s*([•\-*+])\s+/);
```

Comment:
Is it possible to move these outside of the function so you don't need to reinitialize them on each function?